### PR TITLE
[Feature] Governor Bravo

### PR DESCRIPTION
**Pull request by bitparadigm**
1. Adds GovernorBravo contracts from https://github.com/compound-finance/compound-protocol/tree/master/contracts/Governance into repo. (without G1 version which uses V1 storage rather than newer V2).
2. Updates them to 0.8.x version. 
3. Moves Compound Governor contracts to a dedicated folder.
4. Tests GovernorBravo flow.

### CODE REVIEW

File: test/GovernorBravo.test.ts [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+                this.bit.address,
+                this.admin.address,
+                this.governorDelegate.address,
+                PROPOSAL_SPAN,
+                0,
+                VOTING_POWER
+            )
+        )
+        await this.governorDelegator.deployed()
+        // Attach implementation.
```
Comment (CjHare) [inline test/GovernorBravo.test.ts:94]: question: what does this comment mean?

The comment either needs expanding on or deleting.

File: contracts/governor/GovernorBravoInterfaces.sol [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+        address[] targets,
+        uint256[] values,
+        string[] signatures,
+        bytes[] calldatas,
+        uint256 startBlock,
+        uint256 endBlock,
+        string description
+    );
+
+    /// @notice An event emitted when a vote has been cast on a proposal
```
Comment (CjHare) [inline contracts/governor/GovernorBravoInterfaces.sol:19]: suggetsion(blocking): be consistent with multi-line comments, either use `/** ... */` or use `///`

File: contracts/governor/GovernorBravoInterfaces.sol [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+    /// @param reason The reason given for the vote by the voter
+    event VoteCast(
+        address indexed voter,
+        uint256 proposalId,
+        uint8 support,
+        uint256 votes,
+        string reason
+    );
+
+    /// @notice An event emitted when a proposal has been canceled
```
Comment (CjHare) [inline contracts/governor/GovernorBravoInterfaces.sol:33]: question: why use the Doxygen tag on the `event` elements, I didn't think they were picked up on them?

**Review by CjHare — COMMENTED** [review]
Just some questions about the comments and use of Doxygen tags, otherwise looks fine 👍

File: test/GovernorBravo.test.ts [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+            await this.Timelock.deploy(this.admin.address, 1)
+        )
+        await this.timelockAlpha.deployed()
+        this.timelock = <Timelock>(
+            await this.Timelock.deploy(this.admin.address, 1)
+        )
+        await this.timelock.deployed()
+        this.bit = <BitDAO>await this.BitDAO.deploy(this.admin.address)
+        await this.bit.deployed()
+        this.governorAlpha = <GovernorAlpha>(
```
Comment (CjHare) [inline test/GovernorBravo.test.ts:70]: question: Why do you need GovAlpha in a GovBravo test?

Is GovBravo dependent on GovAlpha

**Comment by CjHare** [comment]
question (blocking): Design question; why the copy & paste approach over having Compound Governor Bravo as a dependency and inherit for specialization?

The BitDAO token contract is imported as a dependency from a repo, Bravo could be done the same, with any specialization being done in contract that extends (`is`) it.

There are a number of variations on the approach, all of which would keep the relationship (and feature isolation) clear, while also providing an easy path to merging from upstream.

File: test/GovernorBravo.test.ts (hunk 4 above)
Comment (bitparadigm) [reply test/GovernorBravo.test.ts:70]: As per Compound GovernorBravo is derived from existing GovernorAlpha to make smooth upgrade from it to GovernorBravo as GovernorAlpha had no upgradeability itself and there is easy way to derive GovernorBravo (only need to pull current proposalCount as we didn't change GovernorAlpha state with those proposals except proposalCount).

File: contracts/governor/GovernorBravoInterfaces.sol (hunk 2 above)
Comment (bitparadigm) [reply contracts/governor/GovernorBravoInterfaces.sol:19]: I will replace it, but those comments are from Compound code.

**Comment by bitparadigm** [comment]
> question (blocking): Design question; why the copy & paste approach over having Compound Governor Bravo as a dependency and inherit for specialization?
> 
> The BitDAO token contract is imported as a dependency from a repo, Bravo could be done the same, with any specialization being done in contract that extends (`is`) it.
> 
> There are a number of variations on the approach, all of which would keep the relationship (and feature isolation) clear, while also providing an easy path to merging from upstream.

Shahan requested to have GovernorBravo, but with functionality to have arbitrary vote types, so I will have to upgrade its functionality in next PR, but this requires me having code in our repo, not just importing it. It will be more experimental and not secure contract like native GovernorAlpha rather than proven secure one. 

There might be a good case to just import GovernorAlpha right now as it uses native version right now, but I still did very small alterations like `public` -> `external` in it and replaced Comp with Bit as its really Compound focused implementation with Comp names everywhere even for the token. And GovernorAlpha will not be updated very likely as Compound are focused on developing and updating GovernorBravo.  Other crypto projects don't import GovernorAlpha but do copy paste too.

File: test/GovernorBravo.test.ts (hunk 4 above)
Comment (CjHare) [reply test/GovernorBravo.test.ts:70]: So to be clear, you're using the Alpha for the interface, to ensure Bravo can be interacted with using it?

Comment (bitparadigm) [reply test/GovernorBravo.test.ts:70]: Alpha is need to init Bravo, its not used as in interface. Bravo is continuation from Alpha in literal sense.

**Comment by CjHare** [comment]
>Shahan requested to have GovernorBravo, but with functionality to have arbitrary vote types, so I will have to upgrade its functionality in next PR, but this requires me having code in our repo, not just importing it. It will be more experimental and not secure contract like native GovernorAlpha rather than proven secure one.

After reflecting on this, my opinion is that any specialisation of behaviour should be achieved through inheritance, not copy & paste.
Reasoning being separating the logic being added from the core of Bravo, which allows for easier review of added logic, and maintainability (especially porting from upstream, as the interface is being used rather then being the guts of Bravo, reducing the likelihood of merge conflicts).

If you really need this PR merged in, try convincing @shahankhatch 

Also as this is for experimentation, you could put this into a personal repo, try things out, then port over later. As you're still using git you get the same source control and it's still being developed in public, but allows quicker iterations, as you're not beholden to PR reviews 🙂

File: test/GovernorBravo.test.ts (hunk 4 above)
Comment (shahankhatch) [reply test/GovernorBravo.test.ts:70]: Thanks for pointing out this dependency @bitparadigm. I suggest we continue as is but then adapt Bravo to allow deploying without a prior instance of Alpha.

Comment (bitparadigm) [reply test/GovernorBravo.test.ts:70]: @shahankhatch I wanted to first add GovenorBravo without updates to implementation except version upgrade to 0.8.x and very minor tweaks like comments and things like constants and `external` -> `public`. I can raise an issue to remove reliance on GovernorAlpha to init GovernorBravo and then make a PR specifically for it.

**Comment by shahankhatch** [comment]
In general, I lean towards inheritance for specialization though in this case, I believe we can start with a vanilla copy of Alpha and Bravo. 

We can have the code in our repo since it seems straightforward enough to manage without dealing with different versions at the package management layer, which we might engage with later (as their newer releases of Bravo go through audits).

We can create evolved versions later to allow for better inheritance capabilities without modifying the vanilla versions. 

Although this'll make the governor contracts folder busier with different versions, it'll give us a chance to experiment as we increase the range of gov contracts we wish to support.

**Review by shahankhatch — APPROVED** [review]
lgtm

**Pull request merged**